FIP-113: End Departed Rari Founders vesting of TRIBE

thomas-waite opened the pull request:
# FIP-113: End Departed Rari Founders vesting of TRIBE

Implements the on-chain DAO vote part of https://tribe.fei.money/t/fip-113-end-departed-rari-founders-vesting-of-tribe/4355 

Specifically: 
1. Migrates the Rari infra team timelocks
2. Claws back the vesting TRIBE of Jai Bhavnani and David Lucid

## Timelock migration is performed by:
- Set the `beneficiary` of the existing timelocks to the Tribal Council timelock
- Deploy new Fei and Tribe timelocks for the Rari infra team
- Set the `beneficiary` of the Rari infra team timelocks to be the new Rari ops multisig
- Set the `clawbackAdmin` of the Rari infra team timelocks to be the Tribal Council timelock
- Set the start time as now, the time at which they are deployed
- Set the duration as the remaining duration of the existing timelocks (~580 days)
- Mint the amount of Fei and allocate Tribe to the timelocks as is currently left on the original timelocks

The Rari Infra timelocks are linear, so we can simply mint/allocate remaining funds there and set the duration to the remaining duration.

### PR Checklist - Feature (Proposal)

- [ ] All Tests Passing
- [ ] Proposal Added to ProposalsConfig
- [ ] Fork Block Correct
- [ ] Remove Any .only's on Tests
- [ ] Update Documentation If Needed
- [ ] Update Roles Config
- [ ] Proposal Submitted

thomas-waite committed:
feat: initial commit, accept beneficiaries

thomas-waite committed:
feat: deploy new vesting contracts, implement clawback

thomas-waite committed:
feat: mint FEI and allocate TRIBE to new timelocks

thomas-waite committed:
refactor: clawback tests, template proposal

thomas-waite committed:
feat: deploy and allocate to Lipstone new timelock

thomas-waite committed:
refactor: remove Lipstone vesting

thomas-waite committed:
test: more validation

thomas-waite committed:
test: verify Rari infra beneficiary can claim

thomas-waite committed:
Merge branch 'develop' into clawback

thomas-waite committed:
refactor: switch to setting TC timelock as clawback admin

thomas-waite committed:
refactor: change beneficiary to rariOpsMultisig

thomas-waite committed:
docs: update proposal description

thomas-waite committed:
refactor: remove old beneficiary check

thomas-waite commented on proposals/dao/clawback.ts:
```diff
@@ -0,0 +1,181 @@
+const fipNumber = 'clawback';
+
+// TODO: Update these based on the final balances
+const OLD_RARI_TIMELOCK_FEI_AMOUNT = '3254306506849315068493151';
```
Will update with final numbers once Rari have claimed what is available on their timelocks

thomas-waite replied:
done

thomas-waite commented on protocol-configuration/mainnetAddresses.ts:
```diff
@@ -2092,7 +2097,37 @@ const MainnetContractsConfig: MainnetContractsConfig = {
+    address: '0x4bFa2625D50b68D622D1e71c82ba6Db99BA0d17F',
+    category: AddressCategory.TBD
   }
+  // newRariInfraFeiTimelock: {
```
To fill in when we deploy the new contracts

thomas-waite committed:
refactor: remove executed proposal

eswak commented on proposals/dao/clawback.ts:
```diff
@@ -0,0 +1,183 @@
+  rariTribeTimelockRemainingDuration = await rariInfraTribeTimelock.remainingTime();
+
+  // 1. Deploy new Rari infra vesting contract
+  const LinearTimelockedDelegatorFactory = await ethers.getContractFactory('LinearTimelockedDelegator'); // change timelock type
```
I think you need a 2nd factory to create `LinearTokenTimelock` for the FEI rari infra team vesting contract

thomas-waite replied:
done

eswak commented on proposals/dao/clawback.ts:
```diff
@@ -0,0 +1,183 @@
+
+  // 1. Deploy new Rari infra vesting contract
+  const LinearTimelockedDelegatorFactory = await ethers.getContractFactory('LinearTimelockedDelegator'); // change timelock type
+  const newRariInfraFeiTimelock = await LinearTimelockedDelegatorFactory.deploy(
```
the FEI timelock should be a `LinearTokenTimelock` there is no `delegate(address)` function on the FEI token contract

thomas-waite replied:
Ah good point, oversight

eswak commented on proposals/description/clawback.ts:
```diff
@@ -0,0 +1,57 @@
+      target: 'core',
+      values: '0',
+      method: 'allocateTribe(address,uint256)',
+      arguments: (addresses) => [addresses.newRariInfraTribeTimelock, '3254296867072552004058854'],
```
todo: update with the new balance after rari infra team makes a last claim

thomas-waite replied:
done

eswak commented on proposals/description/clawback.ts:
```diff
@@ -0,0 +1,57 @@
+      target: 'fei',
+      values: '0',
+      method: 'mint(address,uint256)',
+      arguments: (addresses) => [addresses.newRariInfraFeiTimelock, '3254306506849315068493151'],
```
todo: update with the new balance after rari infra team makes a last claim

thomas-waite replied:
done

thomas-waite committed:
refactor: use LinearTokenTimelock for Fei timelock
TimelockedDelegator for Tribe

eswak commented:
just a minor change to do imo (change the factory for Rari infra team FEI's vesting), otherwise lgtm

edit: lol you commited the fix before I commented 💪

thomas-waite committed:
feat: update Fei and Tribe alloc to new Rari balances

ElliotFriedman commented on protocol-configuration/mainnetAddresses.ts:
```diff
@@ -2092,7 +2097,37 @@ const MainnetContractsConfig: MainnetContractsConfig = {
+  //   address: '',
+  //   category: AddressCategory.TBD
+  // },
+  // newRariInfraTribeTimelock: {
```
these should be deployed now so the DAO vote can be submitted

thomas-waite replied:
Getting reviews and approval on this PR first before I deploy

thomas-waite committed:
test: fix unrelated e2es

thomas-waite committed:
deps: update block num to reflect RARI claim

thomas-waite committed:
test: skip rollover tokemak test
forking and fast forwarding doesn't rollover properly
will fix or delete this test in a follow up

thomas-waite committed:
test: remove no longer needed test

eswak reviewed (commented):
lgtm (still need to deploy)

thomas-waite commented on test/integration/tests/fip-38-tokemak.ts:
```diff
@@ -14,7 +14,7 @@ const TOKEMAK_MANAGER_ROLLOVER_ADDRESS = '0x90b6C61B102eA260131aB48377E143D6EB3A
 const IPFS_JSON_FILE_HASH = 'QmP4Vzg45jExr3mcNsx9xxV1fNft95uVzgZGeLtkBXgpkx';
 
-describe('e2e-fip-38-tokemak', function () {
+describe.skip('e2e-fip-38-tokemak', function () {
```
Rolling over doesn't work for some reason when forking and having a proposal that requests a withdrawal. Erwan has the same issue. He's deleted on the release branch as tokemak will go away next week

thomas-waite replied:
Skipping for now, will come back to after this PR if important

thomas-waite commented on test/integration/tests/tribeIncentivesEnd.ts:
```diff
@@ -59,31 +59,7 @@ describe('e2e-end-tribe-incentives', function () {
     await forceEth(curve3LPWhale);
   });
 
-  it('should be able to harvest existing TRIBE rewards and withdraw principle from an LP pool', async () => {
```
No longer relevant/work, proposal to end incentives has finished

Joeysantoro commented on proposals/description/clawback.ts:
```diff
@@ -0,0 +1,57 @@
+
+    ////  Clawbacks
+    {
+      target: 'clawbackVestingContractA',
+      values: '0',
+      method: 'clawback()',
+      arguments: (addresses) => [],
+      description: 'Clawback the TRIBE of Jai Bhavnani and David Lucid'
+    },
+    {
+      target: 'clawbackVestingContractB',
+      values: '0',
+      method: 'clawback()',
+      arguments: (addresses) => [],
+      description: 'Clawback the TRIBE of Jai Bhavnani and David Lucid'
+    },
+    {
+      target: 'clawbackVestingContractC',
+      values: '0',
+      method: 'clawback()',
+      arguments: (addresses) => [],
+      description: 'Clawback the TRIBE of Jai Bhavnani and David Lucid'
+    }
```
the proposal should withdraw all the TRIBE from the timelock back to core after or at least approve some contract like the Tribal Council Timelock to do so

thomas-waite replied:
Have approved the TC timelock for 20M and added checks that it was approved and can transfer

thomas-waite committed:
feat: approve TC timelock for 20M TRIBE from DAO timelock

thomas-waite committed:
test: additional check for approval amount

thomas-waite committed:
test: last check

thomas-waite committed:
build: new Rari infra timelocks

thomas-waite committed:
feat: make use of deployed contracts

thomas-waite committed:
refactor: tidyup deploy script for merging

ElliotFriedman commented on proposals/dao/clawback.ts:
```diff
@@ -0,0 +1,198 @@
+  // const rariInfraTribeTimelock = await ethers.getContractAt(
+  //   'LinearTimelockedDelegator',
+  //   addresses.rariInfraTribeTimelock
+  // );
```
remove this commented code

thomas-waite replied:
Was leaving in in case I needed to do a redeploy (it's needed during deploy time) and will remove after

ElliotFriedman commented on proposals/dao/clawback.ts:
```diff
@@ -0,0 +1,198 @@
+  // );
+
+  // rariFeiTimelockRemainingDuration = await rariInfraFeiTimelock.remainingTime();
+  // rariTribeTimelockRemainingDuration = await rariInfraTribeTimelock.remainingTime();
```
remove commented code

thomas-waite replied:
Was leaving in in case I needed to do a redeploy (it's needed during deploy time) and will remove after
